### DISCUSSION TITLE
piecewise function

### DISCUSSION THREAD

**Discussion by wnn666666** [yalmip/YALMIP#1393] [General discussion]
Hello teacher, how to define the segmentation function in yalmip?
The use implies that the following problem occurs.
F=[F,implies(u1(1),[f1<=f1min,uf1==1])
implies(u1(2),[f1min<=f1<=f1min1.5,uf1==(1.5f1min-f1)/(0.5f1min)])
implies(u1(3),[f1>=f1min1.5,uf1==0])
sum(u1(n,:))==1];
Here u1 is the 0-1binvar variable,f1 is the sdpvar variable,f1min Is a number. uf1 is part of the objective function. When I run, it's not possible to report an error.
But when I set one of the conditions u1(2)==1, it works, why is that?
Why not choose one of the three？

**Comment by johanlofberg** [comment]
I have no idea what you mean with "When I run, it's not possible to report an error"

And please put the code in a code block

**Reply by johanlofberg** [reply]
and you are using u1(1) when you appear to mean u1(n,1) etc

**Comment by wnn666666** [comment]
I'm sorry I didn't make myself clear.Here u1 is the 0-1binvar variable,f1 is the sdpvar variable,f1min Is a number.
The specific code is as follows:

u1=binvar(1,3);
uf1=sdpvar(1,1);
F=[F,sum(u1) == 1];
F=[F,implies(u1(1),[f1<=f1min,uf1==1])];
F=[F,implies(u1(2),[f1min<=f1<=f1min*1.5,uf1==(1.5*f1min-f1)/(0.5*f1min)])];
F=[F,implies(u1(3),[f1>=f1min*1.5,uf1==0])];

uf1 is part of the objective function，However, the running result is not feasible.I just want to express that different f1 ranges cause different uf1 values.

However,When I set one of the conditions to be met

```
F=[F,u1(2)==1];
F=[F,implies(u1(2),[f1min<=f1<=f1min*1.5,uf1==(1.5*f1min-f1)/(0.5*f1min)])];

```
The problem is solvable.         What is the reason for this?

**Reply by johanlofberg** [reply]
You would have to supply a reproducible example, because there is no way in a sensible model that the first would be infeasible and the second feasible

...and again, use code blocks (<> in the menu) to make code readable. I've edited you second block to illustrate

**Comment by wnn666666** [comment]
Sorry about the block I'm not really sure how to output the block.
&nbsp;


This is a screenshot of the code&nbsp; &nbsp;.As I said before, uf1 is part of the objective function and f1min is a number.What exactly is wrong with this piecewise function?


------------------&nbsp;原始邮件&nbsp;------------------
发件人:                                                                                                                        "yalmip/YALMIP"                                                                                    ***@***.***&gt;;
发送时间:&nbsp;2024年4月1日(星期一) 晚上7:49
***@***.***&gt;;
***@***.******@***.***&gt;;
主题:&nbsp;Re: [yalmip/YALMIP] piecewise function (Discussion #1393)





 
You would have to supply a reproducible example, because there is no way in a sensible model that the first would be infeasible and the second feasible
 
...and again, use code blocks (<&gt; in the menu) to make code readable. I've edited you second block to illustrate

**Reply by johanlofberg** [reply]
I cannot run a screenshot...

To put code in a code block, just mark it and press <> in the editor (or look at the code I edited and see what the mark-up is)

**Comment by wnn666666** [comment]
`u1=binvar(1,3);
uf1=sdpvar(1,1);
F=[F,sum(u1) == 1];
F=[F,implies(u1(1),[f1<=f1min,uf1==1])];
F=[F,implies(u1(2),[f1min<=f1<=f1min*1.5,uf1==(1.5*f1min-f1)/(0.5*f1min)])];
F=[F,implies(u1(3),[f1>=f1min*1.5,uf1==0])];`           Is that right？Sorry I didn't see an example of the format.

**Reply by johanlofberg** [reply]
well, you put it in a code block, but you messed up spaces and line-breaks

and that code is not reproducible as I cannot run it

**Comment by wnn666666** [comment]
```
u1=binvar(1,3);
uf1=sdpvar(1,1); 
F=[F,sum(u1) == 1]; 
F=[F,implies(u1(1),[f1<=f1min,uf1==1])]; 
F=[F,implies(u1(2),[f1min<=f1<=f1min*1.5,uf1==(1.5*f1min-f1)/(0.5*f1min)])];
F=[F,implies(u1(3),[f1>=f1min*1.5,uf1==0])];
```

**Reply by wnn666666** [reply]
How about this time

**Reply by johanlofberg** [reply]
as I said, you have to supply reproducible code

**Reply by wnn666666** [reply]
But I've shown here that it's replicable.

**Reply by johanlofberg** [reply]
https://en.wikipedia.org/wiki/Minimal_reproducible_example

**Comment by wnn666666** [comment]
```
u1=binvar(1,3);
uf1=sdpvar(1,1); 
F=[F,sum(u1) == 1]; 
F=[F,implies(u1(1),[f1<=f1min,uf1==1])]; 
F=[F,implies(u1(2),[f1min<=f1<=f1min*1.5,uf1==(1.5*f1min-f1)/(0.5*f1min)])];
F=[F,implies(u1(3),[f1>=f1min*1.5,uf1==0])];
```  
That doesn't work either? I'm sorry I can't get this website open

**Reply by johanlofberg** [reply]
```

>> u1=binvar(1,3);
uf1=sdpvar(1,1); 
F=[F,sum(u1) == 1]; 
F=[F,implies(u1(1),[f1<=f1min,uf1==1])]; 
F=[F,implies(u1(2),[f1min<=f1<=f1min*1.5,uf1==(1.5*f1min-f1)/(0.5*f1min)])];
F=[F,implies(u1(3),[f1>=f1min*1.5,uf1==0])];

Unrecognized function or variable 'F'.
 
>> 
```

**Reply by wnn666666** [reply]
F is the constraint that was defined before.  uf1 here is related to the value range of f1, and uf1 is a part of the objective function, the specific code is as follows.f1 and f2 have expressions, f1min and f2min are just numerical values.
```
F=[];
u1=binvar(1,3);
u2=binvar(1,3);
uf1=sdpvar(1,1); 
uf2=sdpvar(1,1); 
F=[F,sum(u1) == 1]; 
F=[F,implies(u1(1),[f1<=f1min,uf1==1])]; 
F=[F,implies(u1(2),[f1min<=f1<=f1min*1.5,uf1==(1.5*f1min-f1)/(0.5*f1min)])];
F=[F,implies(u1(3),[f1>=f1min*1.5,uf1==0])];
F=[F,sum(u2) == 1]; 
F=[F,implies(u2(1),[f2<=f2min,uf2==1])];      
F=[F,implies(u2(2),[f2min<=f2<=f2min*1.5,uf2==(1.5*f2min-f2)/(0.5*f2min)])];
F=[F,implies(u2(3),[f2>=f2min*1.5,uf2==0])];
 f4=-(0.5*uf1+0.5*uf2);
options=sdpsettings('debug',1,'verbose',2,'solver','gurobi');
result1=optimize(F,f4,options);
```

**Reply by johanlofberg** [reply]
`Unrecognized function or variable 'f1'.`
